### PR TITLE
Inroduce the reduce method for deduplication index building and others 

### PR DESCRIPTION
**Pull request by simcha**
Reduce will return results colleced from each worker. So you need to merge them on the end.

``` Ruby
result = Parallel.reduce(['a','b','c','d','a','b','c','d']) do |result,x|
  result ||= Set.new
  result << x
  result
end
result.compact.reduce(&:+).sort #-> ['a','b','c','d']
```

Remember that if set is small and hi number of workers some workers can return nil. So it is better to use "compact" before merging resuls.

### CODE REVIEW

**Comment by simcha** [comment]
I got no errors on my local run and travis got different errors on my copy of same tree:
https://travis-ci.org/simcha/parallel/builds/168296730.
It seams to be error related to process_diff method
And especially command "ps uaxw|grep ruby|wc -l"
It is not checking only our processes and sometime can result with error.

**Comment by grosser** [comment]
why not map and then reduce after ?

**Comment by simcha** [comment]
My case is word count. I do have 7 million files to count 5 letter words in them. So I need to feed a file and get word count hash. I could map first and have 7 million hashes but merging them is expensive operation and also memory would end. Chunking into smaller packs is not a case as merging ~4 mln. elements hash is as I said expensive. Using the parallel lib with my modification reduced the time of operation of my script 10 fold with very simple change in the code (and 16 CPUs). For that I am very grateful to you for writing the library. Maybe there is some way to do it without a "reduce" method but I do not see it. So answering your question with a question. I do map first but "why do not reduce in parallel" :)

File: Readme.md [hunk 1]
```diff
@@ -37,6 +37,26 @@ items = [1,2,3]
...
 ```
 
+Reduce will return results colleced from each worker. So you need to merge them on the end.
+```Ruby
+result = Parallel.reduce(['a','b','c','d','a','b','c','d']) do |result,x|
+  result ||= Set.new
+  result << x
+  result
+end
+result.compact.reduce(&:+)
```
Comment (grosser) [inline Readme.md]: why does it need compact ?
... maybe do `'x' + 'p'` to show that processing was done

File: Readme.md [hunk 2]
```diff
@@ -37,6 +37,26 @@ items = [1,2,3]
 Parallel.each( -> { items.pop || Parallel::Stop }) { |number| ... }
 ```
 
+Reduce will return results colleced from each worker. So you need to merge them on the end.
```
Comment (grosser) [inline Readme.md]: Maybe write a bit about what you just wrote in the comment ...

```
To avoid overhead from reducing big result sets at the end, use `.reduce` to bla bla
```

**Comment by grosser** [comment]
hmmm would about implementing this with a finish hook ?

```Ruby
>> all = 0; Parallel.each([1,2,3,4,5], finish: -> (_,_,result) { all += result }){ |x| x + 10 }
=> [1, 2, 3, 4, 5]
>> all
=> 65
```

File: Readme.md (hunk 1 above)
Comment (simcha) [inline Readme.md]: I am not absolutely sure I do understand your question but compact is only for a case when processing was not done in some process and it has nil in result. The part of question I do not understand is `'x' + 'p'` part, where to do it?

**Comment by simcha** [comment]
With finish hook how is this parallel? If it runs on each worker finish it will block (?) and it will not give us the overhead reduction if we targeting the case where we replace `all += result` from your example with something longer to run. If it is run in separate workers how you make it so that it returns 1 result and not array of results?

But the finish hook gave me some idea. Now we need to wait for workers to finish to get the result and with the finish hook we can chain calculations. So we can have infinite queue and some time or count parameter to something like finish hook. Every 1 minute or every 100 inputs call the hook with results on input. Seams to me as very simplistic way of creating daemons/servers, what you think?

**Comment by grosser** [comment]
The finish hook is not parallel, but ideally combining results is
fast/simple.

Yep, thats a neat way of doing a parallel daemon or get intermediate
results :)

File: Readme.md (hunk 1 above)
Comment (simcha) [reply Readme.md]: That is fixed now we do not need it.

File: Readme.md (hunk 2 above)
Comment (simcha) [reply Readme.md]: So I added comment similar to that

File: lib/parallel.rb [hunk 3]
```diff
@@ -50,8 +50,14 @@ def initialize(read, write, pid)
     end
 
     def stop
-      close_pipes
+      write.close unless write.closed?
```
Comment (grosser) [inline lib/parallel.rb]: closing multiple times is not an issue, so just do it ?

File: lib/parallel.rb [hunk 4]
```diff
@@ -480,10 +502,13 @@ def extract_count_from_options(options)
       [count, options]
     end
 
-    def call_with_index(item, index, options, &block)
+    def call_with_index(item, index, options, end_result, &block)
```
Comment (grosser) [inline lib/parallel.rb]: arguments after options are weird 🤷‍♂️

File: spec/cases/helper.rb [hunk 5]
```diff
@@ -2,7 +2,7 @@
 require 'parallel'
 
 def process_diff
-  cmd = "ps uaxw|grep ruby|wc -l"
+  cmd = "ps uxw|grep ruby|wc -l"
```
Comment (grosser) [inline spec/cases/helper.rb]: why is this changed ?
can this be it's own PR with a small description ?

File: Readme.md [hunk 6]
```diff
@@ -53,8 +53,27 @@ Parallel.all?([1,2,nil,4,5]) { |number| number != nil }
 # => false
 ```
 
+To avoid overhead from reducing big result sets at the end of processing, use `reduce` this method will return results collected from each worker. You need to merge them on the end to get the fully reduced result.
 
-Processes/Threads are workers, they grab the next piece of work when they finish.
```
Comment (grosser) [inline Readme.md]: only document 1 way of doing things, this is already complicated enough and will only be needed by very few users

**Comment by grosser** [comment]
so basically what this is doing is providing a way to aggregate information per worker without running things on the main thread ?

File: spec/cases/helper.rb (hunk 5 above)
Comment (grosser) [reply spec/cases/helper.rb]: ah found it above.

File: Readme.md (hunk 6 above)
Comment (grosser) [reply Readme.md]: naming it `reduce` suggests that it will return 1 result, but it does not, also we might later add an actual `reduce`
so reduce_by_worker might be more clear

**Comment by grosser** [comment]
also blow up when `isolation` option is used since these 2 make no sense together

**Comment by grosser** [comment]
... also this whole thing can be implemented with some creative hacking too ...

```Ruby
sum = [nil, 0]
Parallel.map [1,2,3,4,5,6,7,8,9], in_processes: 2 do |i|
  sum[0] ||= Process.pid
  sum[1] += i
  sum
end.reverse!.uniq(&:first).map(&:last).sum

```

File: Readme.md (hunk 6 above)
Comment (simcha) [reply Readme.md]: Sure

**Comment by simcha** [comment]
Creative hacking way seam nice
